rust/lance-encoding/src/encodings/physical/bitpack_fastlanes.rs

```diff
@@ -0,0 +1,1547 @@
+                            ($bytes_idx_to_range_indices[i][ranges_idx].end - curr_range_start)
+                                .min(ELEMS_PER_CHUNK - curr_range_start % ELEMS_PER_CHUNK);
+                        decompressed.extend_from_slice(
+                            &decompress_chunk_buf[(curr_range_start % ELEMS_PER_CHUNK) as usize..]
+                                [..elems_this_range_needed_in_this_chunk as usize],
+                        );
+                        if curr_range_start + elems_this_range_needed_in_this_chunk
+                            == $bytes_idx_to_range_indices[i][ranges_idx].end
+                        {
+                            ranges_idx += 1;
+                            if ranges_idx == $bytes_idx_to_range_indices[i].len() {
+                                break;
+                            }
+                            curr_range_start = $bytes_idx_to_range_indices[i][ranges_idx].start;
+                        } else {
+                            curr_range_start += elems_this_range_needed_in_this_chunk;
+                        }
+                    }
+                }
+            }
+        }
+
+        LanceBuffer::reinterpret_vec(decompressed)
+    }};
+}
+
+fn bitpacked_for_non_neg_decode(
+    compressed_bit_width: u64,
+    uncompressed_bits_per_value: u64,
+    data: &[Bytes],
+    bytes_idx_to_range_indices: &[Vec<std::ops::Range<u64>>],
+    num_rows: u64,
+) -> LanceBuffer {
+    match uncompressed_bits_per_value {
+        8 => bitpacked_decode!(
+            u8,
+            compressed_bit_width,
+            data,
+            bytes_idx_to_range_indices,
+            num_rows
+        ),
+        16 => bitpacked_decode!(
+            u16,
+            compressed_bit_width,
+            data,
+            bytes_idx_to_range_indices,
+            num_rows
+        ),
+        32 => bitpacked_decode!(
+            u32,
+            compressed_bit_width,
+            data,
+            bytes_idx_to_range_indices,
+            num_rows
+        ),
+        64 => bitpacked_decode!(
+            u64,
+            compressed_bit_width,
+            data,
+            bytes_idx_to_range_indices,
+            num_rows
+        ),
+        _ => unreachable!(
+            "bitpacked_for_non_neg_decode only supports 8, 16, 32, 64 uncompressed_bits_per_value"
+        ),
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use arrow::array::{
+        Int16Array, Int32Array, Int64Array, Int8Array, UInt16Array, UInt32Array, UInt64Array,
+        UInt8Array,
+    };
+    use arrow::datatypes::DataType;
+
+    #[test_log::test(tokio::test)]
+    async fn test_compute_compressed_bit_width_for_non_neg() {}
+
+    use std::collections::HashMap;
+
+    use lance_datagen::RowCount;
+
+    use crate::testing::{check_round_trip_encoding_of_data, TestCases};
+
+    #[test_log::test(tokio::test)]
+    async fn test_bitpack_fastlanes_u8() {
```

Comment:
I'm not sure these unit tests are using 2.1.  I made a PR to change that.

Reply:
thanks! I will merge it then, 
also, in this PR, the tests in test_bitpack_primitive in `bitpack` will go through `bit_pack_fastlanes`

Reply:
oh my god, these unit tests doesn't go through the codebase at all, fortunately these tests are here before I added the Lance Version 2.1 gate. 

shivers down my spine

merged and thanks so much!